packages/edit-site/src/components/post-list/index.js

```diff
@@ -39,34 +39,57 @@ const { useLocation, useHistory } = unlock( routerPrivateApis );
 const { useEntityRecordsWithPermissions } = unlock( coreDataPrivateApis );
 const EMPTY_ARRAY = [];
 
+const getDefaultView = ( defaultViews, activeView ) =>
+	( defaultViews =
+		defaultViews.find( ( { slug } ) => slug === activeView ).view ||
+		defaultViews[ 0 ].view );
+
 function useView( postType ) {
 	const {
 		params: { activeView = 'all', isCustom = 'false', layout },
 	} = useLocation();
 	const history = useHistory();
-	const DEFAULT_VIEWS = useDefaultViews( { postType } );
-	const selectedDefaultView = useMemo( () => {
-		const defaultView =
-			isCustom === 'false' &&
-			DEFAULT_VIEWS[ postType ].find(
-				( { slug } ) => slug === activeView
-			)?.view;
-		if ( isCustom === 'false' && layout ) {
-			return {
-				...defaultView,
-				type: layout,
-				layout: defaultLayouts[ layout ]?.layout,
-			};
-		}
-		return defaultView;
-	}, [ isCustom, activeView, layout, postType, DEFAULT_VIEWS ] );
-	const [ view, setView ] = useState( selectedDefaultView );
+
+	// PROCESS DEFAULT VIEWS.
+	const defaultViews = useDefaultViews( { postType } );
+	const [ view, setView ] = useState( () =>
+		isCustom === 'false'
+			? getDefaultView( defaultViews, activeView )
+			: defaultViews[ 0 ].view
+	);
+
+	useEffect( () => {
+		setView( ( prevView ) => ( { ...prevView, type: layout } ) );
+	}, [ layout ] );
 
 	useEffect( () => {
-		if ( selectedDefaultView ) {
-			setView( selectedDefaultView );
+		if ( isCustom === 'false' ) {
+			const defaultView = getDefaultView( defaultViews, activeView );
+			setView( defaultView );
+
+			const { params } = history.getLocationWithParams();
+			history.push( {
+				...params,
+				layout: defaultView.type,
```

Comment:
For me ideally, this should not be necessary. I think the "layout" argument in the url should be removed from the links instead.

Reply:
mmm, I'm not sure I follow: don't we need to make sure the layout reflects the view type? Or are you saying that you don't want the layout URL when the view type is list?

Note we need to update the layout when a new default/custom is loaded. This was a source of complexity/bugs: in trunk, the trash view has its own custom hook for this and custom views that use a layout different than `list` don't load properly.

Reply:
I pushed https://github.com/WordPress/gutenberg/pull/63889/commits/6958d2280f3b3da32c9e0a2a3d5bc09795067328 to make it work like trunk.

If there's no layout URL param (first time the Pages page loads) and the view.type is list, the URL param won't be updated. Note that, like trunk, once the layout param is present (user changes layout in dataviews, user changes to a default/custom view with a view.type different than list, etc.), it'll stay there even if the user changes back to list.